### PR TITLE
[opt-tips] Remove -Ounchecked and talk about WMO + 'internal' to reduce dynamic dispatch.

### PR DESCRIPTION
**Pull request by gottesmm**

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@Catfish-Man this one is pretty simple.

File: docs/OptimizationTips.rst [hunk 1]
```diff
@@ -168,6 +164,20 @@ assuming ``E``, ``F`` do not have any overriding declarations in the same file:
     return f.myPrivateVar
   }
 
+Advice: If WMO is enabled, Use 'internal' when declaration does not need to be accessed outside of module
```
Comment (saeta) [inline docs/OptimizationTips.rst]:
```suggestion
Advice: If WMO is enabled, use 'internal' when a declaration does not need to be accessed outside of module
```

**Comment by karwa** [comment]
Is -Ounchecked being removed?!? It’s quite useful for removing things like collection bounds-checking in situations where you can guarantee indexes will always be valid.

**Comment by gottesmm** [comment]
@karwa It is IMO a benchmarking mode that shouldn't really b e used out side of that context. We definitely I don't think want to encourage people to use it.

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by karwa** [comment]
@gottesmm It is documented in lots of places, including the standard library‘s documentation: https://developer.apple.com/documentation/swift/1540960-precondition

Deprecating it was suggested once before, but both Ted and Chris agreed that it would need to go through evolution: https://forums.swift.org/t/deprecating-ounchecked/6928/20

**Comment by gottesmm** [comment]
@karwa well if you notice I am not suggesting that it be deprecated. These are just optimization tips.

**Comment by karwa** [comment]
@gottesmm I understand, but the tips are there for swift users to better understand the available options, right? So given that the mode exists, it makes sense to keep the documentation. Otherwise, users might not understand the caveats and over-use it.

Maybe add a bigger warning if you don’t think the existing one is clear enough, but I don’t see any justification to just not document what it does. It’s a supported feature.

**Comment by gottesmm** [comment]
@karwa This document is not meant to document optimization options. It is meant to provide tips/directions for improving performance when using standard work flows.

Just b/c the project is not deprecating -Ounchecked (where deprecation means giving notice that -Ounchecked will be removed in a future version of swift) doesn't mean that the project should actively push people towards -Ounchecked and suggest people use -Ounchecked. By putting -Ounchecked in this document, we are doing exactly that.

**Pull request merged**